### PR TITLE
Bulk history query

### PR DESCRIPTION
**Pull request by kukukk**

### CODE REVIEW

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:83]: nrOfReplacements -> numberOfReplacements

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:196]: Missing the L from lastCompletedAt

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:242]: Can you replace this log by a more elaborate one ? Something like :
Executing "${adaptedQuery}" with StartTime = ${StartTime.toISOString()} (string formatted),  EndTime = ${EndTime.toISOString()} (string formatted), MaxReturnValues = ${MaxReturnValues}

Displaying only the variable used in the query.

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:265]: We noticed that DateTimeOffset does not support date operation in MSSQL queries. Is there any reason why to use that instead of DateTime ?

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:306]: Need to check how it behaves if endTime is not used but maxReturnValues is used

Comment (burgerni10) [reply src/south/SQLDbToFile/SQLDbToFile.class.js:306]: Same remarks for other divers

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:163]: After we discussed about the migration steps, I think we should allow to use only some of these variables. Particularly StartTime only so it will be easier to automatically migrate LastCompletedDate to StartTime

**Review by burgerni10 — CHANGES_REQUESTED** [review]
I've noticed some conflicts with past rebase of the bulk/develop branch. I'll fix it when merging and I will update the bulk/develop branch after this PR will be merged

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:519]: Can you add specific tests and examples for this method like you did for formatDateWithTimezone method ?

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:503]: Can you add specific tests and examples for this method like you did for formatDateWithTimezone method ?

File: src/south/SQLDbToFile/SQLDbToFile.class.js
Comment (burgerni10) [inline src/south/SQLDbToFile/SQLDbToFile.class.js:188]: Instead of this check, wouldn't it be better to check intervalResult.length AND the last value/timestamp returned from the previous intervalResult ? If both are the same, it will leave the while loop and avoid the problem you saw in PostgreSQL

**Review by burgerni10 — CHANGES_REQUESTED** [review]

**Review by burgerni10 — APPROVED** [review]

**Review by jfhenon — CHANGES_REQUESTED** [review]

File: src/south/SQLDbToFile/SQLDbToFile.schema.jsx
Comment (kukukk) [reply src/south/SQLDbToFile/SQLDbToFile.schema.jsx]: fixed

File: src/south/ProtocolHandler.class.js
Comment (kukukk) [reply src/south/ProtocolHandler.class.js:137]: Indeed, it is not the best naming. `historyQueryImplementation` would be ok?

File: src/south/OPCHDA/OPCHDA.class.js
Comment (kukukk) [reply src/south/OPCHDA/OPCHDA.class.js:66]: For `OPCUA-HA` and `SQLDbToFile` the `historyQuery` implementation returns a promise, so in `ProtolHandler` the `await historyQuery()` call will wait until the response is received. In case of `OPCHDA` the `sendReadMessage` is not returning a promise, so it will immediately return before we receive the response from the agent. The flag for `ongoingReads[scanMode]` is reset after we receive the response from the agent. The `while` loop waits until the response is received.

File: src/south/ProtocolHandler.class.js
Comment (jfhenon) [reply src/south/ProtocolHandler.class.js:137]: Implementation tends to indicate that it is in the child class where you do specific implementation (as it was before). 
Now, it's a part of parentClass so maybe historyQueryHandler or historyQueryHandler

Comment (kukukk) [reply src/south/ProtocolHandler.class.js:137]: historyQueryHandler sounds good.

File: src/south/OPCHDA/OPCHDA.class.js
Comment (jfhenon) [reply src/south/OPCHDA/OPCHDA.class.js:66]: I understand how this works but I would like to rethink this. I think we should await a promise and the response should resolve it. 
However, it's not really related to this PR so we could open a new issue for OPCHDA

File: src/south/ProtocolHandler.class.js
Comment (jfhenon) [inline src/south/ProtocolHandler.class.js:137]: ???

Comment (kukukk) [reply src/south/ProtocolHandler.class.js]: Sorry, I didn't realized that IntelliJ was the active window instead of Terminal when I started to type `git add`. I fixed it.

File: src/south/OPCHDA/OPCHDA.class.js
Comment (kukukk) [reply src/south/OPCHDA/OPCHDA.class.js:66]: Something like this? https://stackoverflow.com/a/36072263/436312

Comment (jfhenon) [reply src/south/OPCHDA/OPCHDA.class.js:66]: yes, seems like a good example. (but let's treat this as another issue)

**Review by burgerni10 — APPROVED** [review]

**Pull request merged**
